addons/l10n_sa_edi/models/account_edi_xml_ubl_21_zatca.py

```diff
@@ -0,0 +1,349 @@
+            :param content: XML content to transform
+            :return: Transformed & Canonicalized XML content
+            """
+            invoice_xsl = etree.parse(get_module_resource('l10n_sa_edi', 'data', 'pre-hash_invoice.xsl'))
+            transform = etree.XSLT(invoice_xsl)
+            return _canonicalize_xml(transform(content))
+
+        root = etree.fromstring(xml_content)
+        # Transform & canonicalize the XML content
+        transformed_xml = _transform_and_canonicalize_xml(root)
+        # Get the SHA256 hashed value of the XML content
+        return sha256(transformed_xml)
+
+    def _l10n_sa_generate_invoice_xml_hash(self, xml_content, mode='hexdigest'):
+        """
+            Generate the b64 encoded sha256 hash of a given xml string:
+                - First: Transform the xml content using a pre-hash_invoice.xsl file
+                - Second: Canonicalize the transformed xml content using the c14n method
+                - Third: hash the canonicalized content using the sha256 algorithm then encode it into b64 format
+        :param str xml_content: The XML content string to be transformed, canonicalized & hashed
+        :param str mode: Name of the function used to return the SHA256 hashing result. Either 'digest' or 'hexdigest'
+        :return: XML content hash
+        :rtype: bytes
+        """
+        xml_sha = self._l10n_sa_generate_invoice_xml_sha(xml_content)
+        if mode == 'hexdigest':
+            xml_hash = xml_sha.hexdigest().encode()
+        elif mode == 'digest':
+            xml_hash = xml_sha.digest()
+        else:
+            raise UserError(_("Only 'hexdigest' and 'digest' methods are supported when generating invoice hash"))
+        return b64encode(xml_hash)
+
+    def _l10n_sa_generate_invoice_hash(self, invoice, mode='hexdigest'):
+        """
+            Function that generates the Base 64 encoded SHA256 hash of a given invoice
+        :param recordset invoice: Invoice to hash
+        :param str mode: Function used to return the SHA256 hashing result. Either 'digest' or 'hexdigest'
+        :return: Given Invoice's hash
+        :rtype: bytes
+        """
+        submission_id = invoice._l10n_sa_get_previous_invoice()
+        if invoice.company_id.l10n_sa_api_mode == 'sandbox' or not submission_id.l10n_sa_attachment_id.datas:
+            # If no invoice, or if using Sandbox, return the b64 encoded SHA256 value of the '0' character
+            return "NWZlY2ViNjZmZmM4NmYzOGQ5NTI3ODZjNmQ2OTZjNzljMmRiYzIzOWRkNGU5MWI0NjcyOWQ3M2EyN2ZiNTdlOQ==".encode()
+        return self._l10n_sa_generate_invoice_xml_hash(b64decode(submission_id.l10n_sa_attachment_id.datas), mode)
+
+    def _get_delivery_vals_list(self, invoice):
+        """
+            Override to include/update values specific to ZATCA's UBL 2.1 specs
+        """
+        res = super()._get_delivery_vals_list(invoice)
+        if 'partner_shipping_id' in invoice._fields:
```

Comment:
no need for this line as partner_shipping_id should always be there.

Reply:
so should this be ignored ? at least for V15 and V14

Reply:
_get_invoice_delivery_partner_id

Reply:
But why do we need the condition because the l10n_sa_delivery_date is always there.